### PR TITLE
fix(influxdb): add org id to permissions

### PR DESCRIPTION
**Pull request by desa**
Closes https://github.com/influxdata/influxdb/issues/10960 

This PR fixes various issues that were introduced by https://github.com/influxdata/platform/pull/2367.

Specifically, the following were done

* `OrgID` was added to permission
* Restructured the string version of a permission was `read:buckets:10` -> `read:orgs/1/buckets/10`
* Allowed does a hierarchical check of permissions `read:orgs/1/buckets` grants `read:org/1/buckets/10` 
* Ensures that authorization permissions `OrgID` match the authorizations `OrgID`.
* Addes `{write,read}-{users,buckets}` flags to CLI.
* Adds `OrgID` to the appropriate tests.

### CODE REVIEW

File: authz.go [hunk 1]
```diff
@@ -40,15 +40,21 @@ func PermissionAllowed(p Permission, ps []Permission) bool {
...
-			permID = *perm.ID
-			if !permID.Valid() {
-				return false
+		if perm.ID == nil {
+			if perm.Action == p.Action && perm.Resource == p.Resource && perm.OrgID == p.OrgID {
+				return true
+			}
+		} else {
+			permID := ID(0)
+			if perm.ID != nil {
```
Comment (goller) [inline authz.go]: Probably simplify as you have already done a nil check

File: bolt/onboarding.go [hunk 2]
```diff
@@ -118,13 +118,12 @@ func (c *Client) Generate(ctx context.Context, req *platform.OnboardingRequest)
 		return nil, err
 	}
 
-	perms := platform.OperPermissions()
-	perms = append(perms, platform.OrgAdminPermissions(o.ID)...)
-	writeBucketPerm, err := platform.NewPermissionAtID(bucket.ID, platform.WriteAction, platform.BucketsResource)
+	perms := platform.OperPermissions(o.ID)
+	writeBucketPerm, err := platform.NewPermissionAtID(bucket.ID, platform.WriteAction, platform.BucketsResource, bucket.OrganizationID)
```
Comment (goller) [inline bolt/onboarding.go]: Shouldn't the onboarding token be for all buckets rather than just a single bucket?

File: bolt/onboarding.go [hunk 3]
```diff
@@ -118,13 +118,12 @@ func (c *Client) Generate(ctx context.Context, req *platform.OnboardingRequest)
...
 	if err != nil {
 		return nil, err
 	}
-	readBucketPerm, err := platform.NewPermissionAtID(bucket.ID, platform.ReadAction, platform.BucketsResource)
+	readBucketPerm, err := platform.NewPermissionAtID(bucket.ID, platform.ReadAction, platform.BucketsResource, bucket.OrganizationID)
```
Comment (goller) [inline bolt/onboarding.go]: perhaps all buckets?

File: inmem/onboarding.go [hunk 4]
```diff
@@ -93,13 +93,12 @@ func (s *Service) Generate(ctx context.Context, req *platform.OnboardingRequest)
 		return nil, err
 	}
 
-	perms := platform.OperPermissions()
-	perms = append(perms, platform.OrgAdminPermissions(o.ID)...)
-	writeBucketPerm, err := platform.NewPermissionAtID(bucket.ID, platform.WriteAction, platform.BucketsResource)
+	perms := platform.OperPermissions(o.ID)
+	writeBucketPerm, err := platform.NewPermissionAtID(bucket.ID, platform.WriteAction, platform.BucketsResource, bucket.OrganizationID)
```
Comment (goller) [inline inmem/onboarding.go]: (same comment... all buckets?)

File: authz.go [hunk 5]
```diff
@@ -40,15 +40,14 @@ func PermissionAllowed(p Permission, ps []Permission) bool {
...
-		if perm.ID != nil {
-			permID = *perm.ID
-			if !permID.Valid() {
-				return false
+		if perm.ID == nil {
+			if perm.Action == p.Action && perm.Resource == p.Resource && perm.OrgID == p.OrgID {
+				return true
+			}
+		} else {
+			if (*perm.ID).Valid() && *perm.ID == pID && perm.Action == p.Action && perm.Resource == p.Resource && perm.OrgID == p.OrgID {
```
Comment (jademcgough) [inline authz.go]: `(*perm.ID).Valid() && *perm.ID == pID`

Won't this always be false? pID is ID(0), which by definition is not valid:
```
// Valid checks whether the receiving ID is a valid one or not.
func (i ID) Valid() bool {
	return i != 0
}
```

Comment (jademcgough) [inline authz.go]: `(*perm.ID).Valid()`

You shouldn't need to check this and `*perm.ID == pID`; we already checked that `pID` is valid on line 37, so anything equal to that must also be valid.

Comment (desa) [reply authz.go]: Good point 👍

File: bolt/onboarding.go (hunk 2 above)
Comment (desa) [reply bolt/onboarding.go]: yup. It is. Dunno why I left this here.

File: bolt/onboarding.go (hunk 3 above)
Comment (desa) [reply bolt/onboarding.go]: yup

**Comment by desa** [comment]
@goller @jademcgough made suggested changes

File: auth.go [hunk 6]
```diff
@@ -23,6 +24,20 @@ type Authorization struct {
 	Permissions []Permission `json:"permissions"`
 }
 
+// Valid ensures that the authorization is valid.
+func (a *Authorization) Valid() error {
+	for _, p := range a.Permissions {
+		if p.OrgID != a.OrgID {
```
Comment (goller) [inline auth.go]: Zero value also works here.  Is there already a check for p.OrgID being valid?

File: authz.go [hunk 7]
```diff
@@ -40,15 +40,14 @@ func PermissionAllowed(p Permission, ps []Permission) bool {
 	}
 
 	for _, perm := range ps {
-		permID := ID(0)
-		if perm.ID != nil {
-			permID = *perm.ID
-			if !permID.Valid() {
-				return false
+		if perm.ID == nil {
```
Comment (goller) [inline authz.go]: What do you think of this to remove else:

```go
if perm.Action != p.Action || perm.Resource != p.Resource || perm.OrgID != p.OrgID {
    continue
}
```

File: auth.go (hunk 6 above)
Comment (desa) [reply auth.go]: There are other checks, but I'll add one here to ensure that the provided orgID is valid.

File: authz.go (hunk 7 above)
Comment (desa) [reply authz.go]: my intuition says keep the explicit if else in this case, but I'm open to switching it if you'd prefer to have it the other way.

**Pull request merged**